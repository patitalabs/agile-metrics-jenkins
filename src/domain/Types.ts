import {CiCollectorConfig, CiMetricItem} from "./ci/Types";

export interface TeamMetricsRequest {
  shouldUpdateEntries: boolean;
  config: any;
}

export interface CoreMetricsService {
  publish(entries: any, shouldReplaceEntries: boolean);
}

export interface CiCollectorService {
  fetch(ciCollectorConfig: CiCollectorConfig): Promise<CiMetricItem[]>;
}
