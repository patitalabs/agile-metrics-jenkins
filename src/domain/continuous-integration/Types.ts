
export interface CiMetricItem {
  jenkinsUrl: string;
  buildName: string;
  result: string;
  durationInSeconds: number;
  causedBy: string;
  revision: string;
  revisionDescription: string;
  buildNumber: string;
  projectName: string;
}

export class CiCollectorConfig  {
  orgName: string;
  projectName: string;
  teamName: string;
  since: Date;
  until?: Date;

  constructor({ orgName, projectName, since, until = null, teamName }) {
    this.orgName = orgName;
    this.projectName = projectName;
    this.since = new Date(since);
    this.until = until ? new Date(until) : null;
    this.teamName = teamName;
  }
}