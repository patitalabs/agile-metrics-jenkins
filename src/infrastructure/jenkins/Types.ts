export class JenkinsJob {
  name: string;
  branch: string;
  builds: JenkinsBuild[];
}

export class JenkinsBuild {
  timestamp: Date;
  result: string;
  durationInMs: number;
  causedBy: string;
  revision: string;
  revisionDescription: string;
  buildNumber: string;
  url: string;
}

export interface JenkinsService {
  findData(
    orgName: string,
    projectName: string,
    since: Date,
    until?: Date
  ): Promise<JenkinsJob>;
}

export interface JenkinsClient {
  getData(url): Promise<any>;
}

export interface JenkinsRepository {
  jobDetails(
    orgName: string,
    projectName,
    since: Date,
    until?: Date
  ): Promise<JenkinsJob>;
}
