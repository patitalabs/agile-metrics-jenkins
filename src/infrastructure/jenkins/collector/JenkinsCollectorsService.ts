import { JenkinsMetricConverter } from './JenkinsMetricConverter';
import { JenkinsService } from '../Types';
import {CiCollectorConfig, CiMetricItem} from "../../../domain/ci/Types";

export class JenkinsCollectorsService {
  constructor(private readonly jenkinsService: JenkinsService) {}

  public async fetch(
    jenkinsConfig: CiCollectorConfig
  ): Promise<CiMetricItem[]> {
    const jenkinsJob = await this.jenkinsService.findData(
      jenkinsConfig.orgName,
      jenkinsConfig.projectName,
      jenkinsConfig.since,
      jenkinsConfig.until
    );

    return JenkinsMetricConverter.toMetricItem(jenkinsJob, jenkinsConfig);
  }
}
