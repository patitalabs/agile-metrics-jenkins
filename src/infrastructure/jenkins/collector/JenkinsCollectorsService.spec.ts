import { JenkinsCollectorsService } from './JenkinsCollectorsService';
import { JenkinsJob, JenkinsService } from '../Types';
import {CiCollectorConfig} from "../../../domain/continuous-integration/Types";

function testJenkinsJob(): JenkinsJob {
  return {
    name: 'jenkins job',
    branch: 'master',
    builds: [
      {
        timestamp: new Date('2018-12-06'),
        result: 'SUCCESS',
        durationInMs: 2000,
        causedBy: 'someone',
        revision: 'someVersion',
        revisionDescription: 'some description',
        buildNumber: '50',
        url: 'someUrl',
      },
    ],
  };
}

describe('JenkinsCollectorsService', () => {
  const jenkinsService: JenkinsService = {
    findData: (
      orgName: string,
      projectName: string,
      since: Date
    ): Promise<JenkinsJob> => Promise.resolve(testJenkinsJob()),
  };

  const jenkinsCollectorsService: JenkinsCollectorsService = new JenkinsCollectorsService(
    jenkinsService
  );

  it('should fetch jenkinsMetrics', async () => {
    const ciCollectorConfig: CiCollectorConfig = new CiCollectorConfig(
      {
        orgName: 'orgName',
        teamName: 'someTeamName',
        projectName: 'projectName',
        since: '2018-11-20',
        until: '2020-11-20',
      }
    );

    const data = await jenkinsCollectorsService.fetch(ciCollectorConfig);
    expect(data).toMatchSnapshot();
  });
});
