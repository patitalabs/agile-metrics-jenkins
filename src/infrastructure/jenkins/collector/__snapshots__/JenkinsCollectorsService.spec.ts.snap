// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JenkinsCollectorsService should fetch jenkinsMetrics 1`] = `
Array [
  Object {
    "buildName": "jenkins job",
    "buildNumber": "50",
    "causedBy": "someone",
    "createdAt": 2018-12-06T00:00:00.000Z,
    "dataType": "CI",
    "durationInSeconds": 2,
    "id": "c98207bf100109f4109cc33b99b467a6ce42299ff4f1fc8d33784ca262550a6d",
    "jenkinsUrl": "someUrl",
    "projectName": "projectName",
    "result": "SUCCESS",
    "revision": "someVersion",
    "revisionDescription": "some description",
    "teamName": "someTeamName",
  },
]
`;
